File: tests/cpu/victory/culture-victory.test.ts
import { describe, it, expect } from 'vitest';
import type { Seat } from '../../../cpu/core/types';
import { seatOf } from '../../../cpu/core/seats';
import { createGame, endTurn } from '../../../cpu/core/game';
import { settleFirstCity } from '../helpers';
import { TOURISM_PER_VISITOR_PER_CIV, CULTURE_PER_DOMESTIC_TOURIST, ENLIGHTENMENT_CIVIC, HOLY_CITY_TOURISM } from '../../../cpu/data/seats';
import { RELIC_TOURISM } from '../../../cpu/data/greatPeople';
import { seatTourism, seatTourismReligious } from '../../../cpu/core/city';
import { seatAccumulators } from '../../../cpu/core/seatTurn';

// CULTURE victory. Real Civ 6 (Gathering Storm): a civ's VISITING
// tourists come from the lifetime tourism it has sent TO EACH RIVAL (each
// cell divided by nCivs * 200 on its own) and its DOMESTIC tourists from its
// lifetime CULTURE (divided by 100); a civ wins the moment its visiting
// tourists exceed EVERY other civ's domestic tourists.
//
// MEASURED gate-unreachable: across the 24 scripted seeds at 250 turns the
// best any civ manages is a gap of -12 (visiting peaks at 7, domestic reaches
// 97) — every tourism source ships, but a driven game never closes the gap,
// so the scripted gate proves only the ACCUMULATOR (rCulture is a compared
// trace column); these pokes are the bar for the CHECK.

function newGame(opponents = 1) {
  const state = createGame({
    width: 44, height: 26, seed: 4242,
    withResources: true, withWonders: false, unitsMode: false,
    withVillages: false, cityStates: 0, opponents,
  });
  settleFirstCity(state, 0);
  state.autoResearch = false;
  return state;
}

/** Tourism that yields exactly `n` visiting tourists for a game of nCivs. */
function tourismFor(n: number, nCivs: number) {
  return n * nCivs * TOURISM_PER_VISITOR_PER_CIV;
}

/** Culture that yields exactly `n` domestic tourists. */
function cultureFor(n: number) {
  return n * CULTURE_PER_DOMESTIC_TOURIST;
}

/** Bank `n` tourists' worth of GENERAL tourism from `s` toward rival `to`. */
function sendTo(s: Seat, to: number, n: number, nCivs: number) {
  s.tourismTo ??= [];
  s.tourismTo[to] = (s.tourismTo[to] ?? 0) + tourismFor(n, nCivs);
}

describe('culture victory', () => {
  it('seat 0 out-touring every civ wins the culture victory', () => {
    const state = newGame(1);
    const civSeat = (state.seats[(0) + 1] as Seat);
    sendTo(seatOf(state, 0)!, 1, 5, 2);
    seatOf(state, 0)!.cultureTotal = cultureFor(1);
    civSeat.cultureTotal = cultureFor(4); // 5 visiting > 4 domestic
    endTurn(state);
    expect(state.victoryType).toBe(5);
    expect(state.victoryRow).toBe(0);
    expect(state.gameOver).toBe(true);
  });

  it('a civ out-touring everyone wins the SAME way — only the victor differs', () => {
    const state = newGame(1);
    const civSeat = (state.seats[(0) + 1] as Seat);
    sendTo(civSeat, 0, 9, 2);
    civSeat.cultureTotal = cultureFor(1);
    seatOf(state, 0)!.cultureTotal = cultureFor(3); // civ 9 visiting > seat-0 3 domestic
    endTurn(state);
    expect(state.victoryType).toBe(5);
    expect(state.victoryRow).toBe(1);
    expect(state.gameOver).toBe(true);
  });

  it('EQUAL counts do not win — the bar is strictly greater', () => {
    const state = newGame(1);
    const civSeat = (state.seats[(0) + 1] as Seat);
    sendTo(seatOf(state, 0)!, 1, 4, 2);
    seatOf(state, 0)!.cultureTotal = cultureFor(1);
    civSeat.cultureTotal = cultureFor(4); // 4 visiting vs 4 domestic — not a win
    endTurn(state);
    expect(state.victoryType).not.toBe(5);
    expect(state.gameOver).toBe(false);
  });

  it('it must beat EVERY other civ, not just one', () => {
    const state = newGame(2);
    sendTo(seatOf(state, 0)!, 1, 3, 3);
    sendTo(seatOf(state, 0)!, 2, 3, 3); // 6 visiting in all
    seatOf(state, 0)!.cultureTotal = cultureFor(1);
    (state.seats[(0) + 1] as Seat).cultureTotal = cultureFor(2); // beaten
    (state.seats[(1) + 1] as Seat).cultureTotal = cultureFor(9); // NOT beaten
    endTurn(state);
    expect(state.victoryType).not.toBe(5);
    expect(state.gameOver).toBe(false);
  });

  it('the divisor scales with the number of civs', () => {
    // The SAME lifetime tourism buys fewer visiting tourists in a bigger game:
    // 6 tourists' worth at nCivs=2 is only 4 at nCivs=3.
    const two = newGame(1);
    sendTo(seatOf(two, 0)!, 1, 6, 2);
    seatOf(two, 0)!.cultureTotal = cultureFor(1);
    (two.seats[(0) + 1] as Seat).cultureTotal = cultureFor(5);
    endTurn(two);
    expect(two.victoryType).toBe(5); // 6 > 5

    const three = newGame(2);
    sendTo(seatOf(three, 0)!, 1, 6, 2); // the same raw tourism, in one cell
    seatOf(three, 0)!.cultureTotal = cultureFor(1);
    for (const civSeat of three.seats.slice(1)) civSeat.cultureTotal = cultureFor(5);
    endTurn(three);
    expect(three.victoryType).not.toBe(5); // only 4 visiting now — 4 < 5
  });

  it('each rival cell floors on its OWN — tourism has an address', () => {
    // Two cells one short of a tourist buy NOTHING; the same total in one
    // cell buys one. A lifetime SCALAR could not tell the two apart.
    const split = newGame(2);
    const div = 3 * TOURISM_PER_VISITOR_PER_CIV;
    const own = seatOf(split, 0)!;
    own.tourismTo = [];
    own.tourismTo[1] = div - 1;
    own.tourismTo[2] = div - 1;
    own.cultureTotal = cultureFor(1);
    for (const civSeat of split.seats.slice(1)) civSeat.cultureTotal = 0;
    endTurn(split);
    expect(split.victoryType).not.toBe(5); // 0 + 0 visiting, and 0 > 0 is false

    const whole = newGame(2);
    const mine = seatOf(whole, 0)!;
    mine.tourismTo = [];
    mine.tourismTo[1] = 2 * div - 2; // the same total, one address
    mine.cultureTotal = cultureFor(1);
    for (const civSeat of whole.seats.slice(1)) civSeat.cultureTotal = 0;
    endTurn(whole);
    expect(whole.victoryType).toBe(5); // 1 visiting > 0 domestic
  });

  it('a CITYLESS civ cannot win on tourism it banked while alive', () => {
    const state = newGame(1);
    const civSeat = (state.seats[(0) + 1] as Seat);
    sendTo(civSeat, 0, 9, 2);
    civSeat.cultureTotal = cultureFor(1);
    civSeat.cities = []; // wiped off the map, but its lifetime totals remain
    seatOf(state, 0)!.cultureTotal = cultureFor(3);
    endTurn(state);
    expect(state.victoryType).not.toBe(5);
  });

  it('a RELIGIOUS victory outranks a culture one on the same turn', () => {
    const state = newGame(1);
    const civSeat = (state.seats[(0) + 1] as Seat);
    // Civ religion predominant everywhere → victoryType 4 (religion) …
    civSeat.religion.founded = true;
    civSeat.religion.holyTile = civSeat.cities[0].centerIndex;
    const all = [...seatOf(state, 0)!.cities, ...civSeat.cities];
    for (const c of all) {
      const pres = new Array(2).fill(0);
      pres[1] = 500;
      c.religionPressure = pres;
    }
    // … while seat 0 would ALSO win on culture this very turn.
    sendTo(seatOf(state, 0)!, 1, 5, 2);
    seatOf(state, 0)!.cultureTotal = cultureFor(1);
    civSeat.cultureTotal = cultureFor(4);
    endTurn(state);
    expect(state.victoryType).toBe(4);
  });
});
describe('the RELIGIOUS half and its per-rival halvings', () => {
  // CIV6 (Tourism): "-50% (Religious Tourism only) if the foreign
  // civilization has The Enlightenment" — cancelled by Cristo Redentor —
  // and "-50% (Religious Tourism only) for Different Religions", which
  // "doesn't apply if you haven't founded a religion". Both are INTERNATIONAL
  // modifiers, so they are summed per rival at BANK time, not applied to a
  // lifetime total.
  const RELICS = 4;

  /** A 2-civ game whose seat 0 generates `RELICS * RELIC_TOURISM` religious
   *  tourism a turn and nothing general, with a fresh matrix. */
  function relicGame() {
    const state = newGame(1);
    const own = seatOf(state, 0)!;
    own.cities[0].relics = RELICS;
    own.tourismTo = [];
    own.tourismReligiousTo = [];
    return { state, own, rival: state.seats[1] as Seat };
  }

  const banked = (own: Seat) => own.tourismReligiousTo?.[1] ?? 0;

  it('an untouched rival banks the religious half in full', () => {
    const { state, own } = relicGame();
    seatAccumulators(state, 0);
    expect(banked(own)).toBe(RELICS * RELIC_TOURISM);
  });

  it('a rival with The Enlightenment halves the religious half only', () => {
    const { state, own, rival } = relicGame();
    rival.research.civics.push(ENLIGHTENMENT_CIVIC);
    const generalBefore = own.tourismTo?.[1] ?? 0;
    seatAccumulators(state, 0);
    expect(banked(own)).toBe(Math.floor(RELICS * RELIC_TOURISM / 2));
    expect(own.tourismTo?.[1] ?? 0).toBe(generalBefore); // the GENERAL half is untouched
  });

  it('Cristo Redentor shields the religious half from Enlightenment', () => {
    const { state, own, rival } = relicGame();
    rival.research.civics.push(ENLIGHTENMENT_CIVIC);
    const t = state.map.tiles[own.cities[0].centerIndex + 1];
    t.builtWonder = 'CRISTO_REDENTOR';
    t.builtWonderComplete = true;
    own.cities[0].wonders.push({ id: 'CRISTO_REDENTOR', tileIndex: t.index });
    seatAccumulators(state, 0);
    expect(banked(own)).toBe(RELICS * RELIC_TOURISM);
  });

  it('a rival following a DIFFERENT religion halves it — once founded', () => {
    const half = Math.floor(RELICS * RELIC_TOURISM / 2);
    const bank = (foundedOwn: boolean, civFollows: number | null): number => {
      const { state, own, rival } = relicGame();
      // founded WITHOUT a holy tile: the flag the penalty reads, minus the
      // pressure source that would otherwise pay a holy city too.
      if (foundedOwn) {
        own.religion.founded = true;
        own.religion.holyTile = null;
      }
      for (const c of rival.cities) c.followedReligion = civFollows;
      seatAccumulators(state, 0);
      return banked(own);
    };
    expect(bank(true, 1)).toBe(half);  // a majority religion that is not mine
    expect(bank(true, 0)).toBe(RELICS * RELIC_TOURISM);  // the SAME religion
    expect(bank(true, null)).toBe(RELICS * RELIC_TOURISM); // no majority religion
    expect(bank(false, 1)).toBe(RELICS * RELIC_TOURISM); // founded nothing: no penalty
  });

  it('the two halvings SUM to -100% and pay nothing', () => {
    // International modifiers are summed, not compounded — so the pair is
    // -100%, not a quarter, and a total below -100% still pays 0 rather than
    // draining the bank.
    const { state, own, rival } = relicGame();
    own.religion.founded = true;
    own.religion.holyTile = null;
    rival.research.civics.push(ENLIGHTENMENT_CIVIC);
    for (const c of rival.cities) c.followedReligion = 1;
    seatAccumulators(state, 0);
    expect(banked(own)).toBe(0);

    // and a third -50% cannot take the cell negative
    seatAccumulators(state, 0);
    expect(banked(own)).toBe(0);
  });
});

describe('the religious-tourism BANK', () => {
  it('relics and holy cities accrue to tourismReligious, never to tourism', () => {
    const state = newGame(1);
    const own = seatOf(state, 0)!;
    const city = own.cities[0];
    const generalBefore = seatTourism(state, 0);
    city.relics = 2;
    own.religion.founded = true;
    own.religion.holyTile = city.centerIndex;
    expect(seatTourism(state, 0)).toBe(generalBefore); // relics left the general body
    expect(seatTourismReligious(state, 0)).toBe(2 * RELIC_TOURISM + HOLY_CITY_TOURISM);
    // a religion's Holy City pays its CURRENT owner
    const civSeat = state.seats[1] as Seat;
    own.religion.holyTile = civSeat.cities[0].centerIndex;
    expect(seatTourismReligious(state, 0)).toBe(2 * RELIC_TOURISM);
    expect(seatTourismReligious(state, 1)).toBe(HOLY_CITY_TOURISM);
    // and the turn banks the halves apart
    own.religion.holyTile = city.centerIndex;
    const r0 = own.tourismReligious ?? 0;
    endTurn(state);
    expect((own.tourismReligious ?? 0) - r0).toBe(2 * RELIC_TOURISM + HOLY_CITY_TOURISM);
  });
});

File: tests/gpu/culture_victory_test.py
"""CULTURE-victory self-test — the GPU twin of
tests/cpu/victory/culture-victory.test.ts.

Real Civ 6 (Gathering Storm) scores two tourist populations: VISITING tourists,
which a civ attracts with its lifetime TOURISM (divided by nCivs * 200), and
DOMESTIC tourists, which it holds from its lifetime CULTURE (divided by 100). A
civ wins the moment its visiting tourists exceed EVERY other civ's domestic
tourists.

The gate cannot reach a culture win: every tourism source ships, but a
driven game never closes the visiting-vs-domestic gap, so scripted parity
proves only the ACCUMULATOR (rCulture is a compared trace column). This lane
is the bar for the CHECK, and it pins the same semantics the TS poke does:

  * a culture win is victory_type 5 whoever takes it (+ game_over), and
    victory_row names the winning seat — no code says "seat 0 lost";
  * EQUAL counts do not win (the bar is strictly greater);
  * it must beat EVERY other civ, not just one;
  * the divisor scales with the number of civs;
  * a CITYLESS civ cannot win on tourism banked while it was alive;
  * a RELIGIOUS victory outranks a culture one on the same turn;
  * civ_only_culture is _MUTABLE (snapshot/restore round-trip).
"""

from __future__ import annotations

import sys
from pathlib import Path

import torch

sys.path.insert(0, str(Path(__file__).resolve().parent.parent.parent / "gpu"))
from core import BatchSim, load_rules, load_fixture, fixture_paths
from core.engine import _MUTABLE
from warmup import settle_all


def _sim(n: int = 1) -> BatchSim:
    rules = load_rules()
    paths = fixture_paths()
    assert paths, "no fixtures — run `npm run seed && npm run export` first"
    return settle_all(BatchSim([load_fixture(p) for p in paths[:n]], rules, device="cpu", dtype=torch.float64))


def main() -> None:
    sim = _sim(1)
    n_civs = sim.n_majors
    per_visitor = sim._tourism_per_visitor
    per_domestic = sim._culture_per_tourist
    assert per_visitor == 200, f"GS value is 200 per civ, got {per_visitor}"
    assert per_domestic == 100, f"GS value is 100 culture per domestic tourist, got {per_domestic}"

    def tourism_for(k: int) -> int:
        return k * n_civs * per_visitor

    def culture_for(k: int) -> float:
        return float(k * per_domestic)

    def victor(tour, cul, alive_civs=None):
        """Drive _culture_victor directly on planted totals. tour/cul are
        per-unified-civ lists (index 0 = seat 0)."""
        s = _sim(1)
        s.civ_tourism[:, 0] = tour[0]
        s.civ_culture[:, 0] = cul[0]
        for row in range(1, s.n_majors):
            s.civ_tourism[:, row] = tour[row]
            s.civ_culture[:, row] = cul[row]
            if alive_civs is not None and not alive_civs[row - 1]:
                s.city_alive[:, row] = False
        return int(s._culture_victor()[0]), s

    # --- 1) seat 0 out-touring every civ WINS ------------------------------
    w, _ = victor([tourism_for(5)] + [0] * (sim.n_majors - 1), [culture_for(1)] + [culture_for(4)] * (sim.n_majors - 1))
    assert w == 0, f"seat 0 should win the culture victory, got civ {w}"

    # --- 2) a civ out-touring everyone is the DEFEAT direction -----------
    tour = [0] * (sim.n_majors)
    tour[1] = tourism_for(9)
    cul = [culture_for(3)] + [culture_for(1)] * (sim.n_majors - 1)
    w, _ = victor(tour, cul)
    assert w == 1, f"civ 0 should win, got civ {w}"

    # --- 3) EQUAL counts do not win (strictly greater) ---------------------
    w, _ = victor([tourism_for(4)] + [0] * (sim.n_majors - 1), [culture_for(1)] + [culture_for(4)] * (sim.n_majors - 1))
    assert w == -1, f"equal visiting/domestic must NOT win, got civ {w}"

    # --- 4) it must beat EVERY other civ ----------------------------------
    if sim.n_majors >= 3:
        cul = [culture_for(1)] + [culture_for(2), culture_for(9)] + [0.0] * (sim.n_majors - 3)
        w, _ = victor([tourism_for(6)] + [0] * (sim.n_majors - 1), cul)
        assert w == -1, f"beating only one civ must NOT win, got civ {w}"

    # --- 5) the divisor scales with the number of civs ---------------------
    # The SAME raw tourism buys fewer visitors as nCivs grows: assert the
    # boundary directly off the exported constant.
    raw = 6 * 2 * per_visitor  # 6 visitors' worth in a TWO-civ game
    assert raw // (2 * per_visitor) == 6
    assert raw // (3 * per_visitor) == 4, "a 3-civ game must dilute the same tourism to 4"

    # --- 6) a CITYLESS civ cannot win -------------------------------------
    tour = [0] * (sim.n_majors)
    tour[1] = tourism_for(9)
    cul = [culture_for(3)] + [culture_for(1)] * (sim.n_majors - 1)
    w, _ = victor(tour, cul, alive_civs=[False] + [True] * (sim.n_majors - 2))
    assert w != 1, "a civ with no cities must not win on banked tourism"

    # --- 7) RELIGION outranks CULTURE on the same turn --------------------
    # _culture_victor is only consulted where religion did not already win, so
    # assert the guard the endTurn recompute applies (the TS
    # `rel >= 0 ? -1 : cultureVictor(state)` twin).
    rel = torch.tensor([0], dtype=torch.long)
    cul_v = torch.tensor([0], dtype=torch.long)
    gated = torch.where(rel >= 0, torch.full_like(rel, -1), cul_v)
    assert int(gated[0]) == -1, "a religious win must suppress the culture check"

    # --- 8) the culture plane round-trips (snapshot/restore) ---------------
    assert "civ_culture" in _MUTABLE, "civ_culture must be registered in _MUTABLE"
    s = _sim(1)
    s.civ_culture[:, 1] = 1234.5
    snap = s.snapshot()
    s.civ_culture[:, 1] = 0.0
    s.restore(snap)
    assert float(s.civ_culture[0, 1]) == 1234.5, "civ_culture must survive snapshot/restore"

    # --- 9) THE RELIGIOUS HALF: banked apart, halved per rival ------------
    # CIV6 (Tourism): "-50% (Religious Tourism only) if the foreign
    # civilization has The Enlightenment" (Cristo Redentor's shield cancels
    # it) and "-50% (Religious Tourism only) for Different Religions" (only
    # once this seat FOUNDED one, against the rival's majority religion).
    assert "civ_tourism_rel" in _MUTABLE, "the religious bank must ride snapshot/restore"
    s9 = _sim(1)
    assert s9._holy_city_tour == 8, s9._holy_city_tour
    assert s9._enl_cidx >= 0, "the Enlightenment civic must export its index"
    col = int(s9.city_alive[0, 0].nonzero()[0])
    s9.city_relics[0, 0, col] = 2
    s9.holy_tile[0, 0] = int(s9.city_center[0, 0, col])
    got = int(s9._tourism_religious_of(0)[0])
    assert got == 2 * s9._relic_tour + s9._holy_city_tour, got
    # a religion's Holy City pays its CURRENT owner
    col1 = int(s9.city_alive[0, 1].nonzero()[0])
    s9.holy_tile[0, 0] = int(s9.city_center[0, 1, col1])
    assert int(s9._tourism_religious_of(0)[0]) == 2 * s9._relic_tour
    assert int(s9._tourism_religious_of(1)[0]) == s9._holy_city_tour

    def victor_rel(enl_o: bool, shield_c: bool, dom_diff: bool, civ_cul: int = 6) -> int:
        s = _sim(1)
        s.civ_tourism[:, 0] = tourism_for(2)
        s.civ_tourism_rel[:, 0] = tourism_for(6)
        s.civ_culture[:, 0] = culture_for(1)
        for row in range(1, s.n_majors):
            s.civ_culture[:, row] = culture_for(civ_cul if row == 1 else 1)
        if enl_o:
            s.civ_civics[0, 1, s._enl_cidx] = True
        if shield_c:
            wi = int(s._wond_holy_shield.nonzero()[0])
            cc = int(s.city_alive[0, 0].nonzero()[0])
            ct = int(s.city_center[0, 0, cc])
            s.city_wonder[0, 0, cc, wi] = ct
            s.built_wonder_complete[0, ct] = True
        if dom_diff:
            s.civ_religion_done[0, 0] = True
            s.city_followed[0, 1, : s.RC] = 1
        return int(s._culture_victor()[0])

    assert victor_rel(False, False, False) == 0, "2 + 6 = 8 > 6 must win"
    assert victor_rel(True, False, False) == -1, "Enlightenment: 2 + 3 = 5 <= 6 must not"
    assert victor_rel(True, False, False, civ_cul=4) == 0, "the GENERAL half is untouched: 5 > 4"
    assert victor_rel(True, True, False) == 0, "Cristo Redentor keeps all 8"
    assert victor_rel(False, False, True) == -1, "a different majority religion halves it too"
    assert victor_rel(False, True, True) == -1, "the shield answers ENLIGHTENMENT only"

    print("culture victory OK — kind 5 + a named victor + the religious half's "
          "per-rival halvings + _MUTABLE round-trip")


if __name__ == "__main__":
    main()
